modin/test/storage_formats/pandas/test_internals.py

```diff
@@ -110,3 +110,38 @@ def func_to_apply(partition, row_internal_indices, col_internal_indices, item):
     md_df._query_compiler._modin_frame = new_modin_frame
 
     df_equals(md_df, pd_df)
+
+
+@pytest.mark.skipif(
+    Engine.get() != "Dask" and Engine.get() != "Ray",
+    reason="Rebalancing partitions is only supported for Dask and Ray engines",
+)
+def test_rebalance_partitions():
+    small_dfs = [
+        pd.DataFrame(
+            [[i + j for j in range(0, 100)]],
+            columns=[f"col{j}" for j in range(1, 101)],
+            index=pd.Index([i - 1]),
+        )
+        for i in range(1, 10001, 100)
+    ]
+    large_df = pd.concat(small_dfs)
+    large_modin_frame = large_df._query_compiler._modin_frame
+    assert large_modin_frame._partitions.shape == (
+        4,
+        4,
+    ), "Partitions were not rebalanced after concat."
+    assert all(
+        isinstance(ptn, large_modin_frame._partition_mgr_cls._column_partitions_class)
+        for ptn in large_modin_frame._partitions.flatten()
+    )
+    large_df = large_df.apply(lambda x: x + 1)
```

Comment:
Could you please add a comment about why you're testing what the apply does? One reason I'm thinking of is that it requires building full-axis virtual partitions out of non-full-axis virtual partitions.

Reply:
Isn't this a full-axis apply? I thought we would create full-axis virtual partitions out the virtual partitions, apply the UDF to the full-axis partitions, then split the virtual partitions into block partitions?

Reply:
You are correct - my bad!